[Consolidated Channel] Update target pod IPs when dispatcher pods scale before the subscription is marked ready

devguyio opened the pull request:
## Background
The existing dispatcher pods Prober acts like this:
1. `IsReady(sub)` -> get list of target pods, cache the state (including the list of pods), and start probing those pods.
3. Probing collects ready partitions and stores them in the cached target state
4. After each dispatcher pod partitions are cached, check if we collected all partitions we are looking for and if so, mark the cached targetState as ready.  

There's a corner case the manifests itself as follows:
1. `IsReady(sub)` -> get list of target pods, cache the state (including the list of pods), and start probing those pods.
2.  Dispatcher pods scale up before the cached targetState becomes ready. 
3.  No new probes are created for the new pods.
4. The existing probing returns only a subset of the partitions, because not all dispatcher pods are probed for this targetState.

This is a fix for the behavior [described here](https://github.com/knative-sandbox/eventing-kafka/pull/476#issuecomment-809156584)

## Proposed Changes
### Kafka Controller 
1. Refresh prober pods when dispatcher pods are added or deleted.

### Prober
1. Refactor the code to be more readible by breaking down `IsReady` and getting rid of non-readable anonymous function and logic.
2. Consider an in-flight targetState outdated if the list of initial probed pods change from the current, forget the state and reprobe new pods.
3. Add `RefreshPodProbing()` that will eject outdated in-flight target states, and reprobe the new pods instead.

devguyio committed:
Restart probing dispatchers if the pods change before being ready

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
fix when to return subscribers reconciled

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Proper locking in status prober

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio committed:
Add a debug log when dispatcher pods change

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

lberk commented on pkg/channel/consolidated/reconciler/controller/kafkachannel.go:
```diff
@@ -225,34 +230,30 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, kc *v1beta1.KafkaChannel
-	if err != nil {
-		logger.Errorw("Error closing the connection", zap.Error(err))
-		return err
+	if !reconciled {
```
why wouldn't this be considered a new error type and then we can warn on that?

devguyio replied:
can you explain more? as in consider the case where not all subscribers were reconciled an error and check the error returned by `reconcileSubscribers` instead of a bool and error?

lberk replied:
right.  We're mixing the return value across multiple variables at this point.  For example, the fact we sometimes return `reconciled` and othertimes elect to return a constant (`false`) indicates to me, that something is amiss with the logic as it currently stands.

let's keep the single return value (of type `error`) and have the control flow dependant on the actual error.

lberk commented on pkg/channel/consolidated/reconciler/controller/kafkachannel.go:
```diff
@@ -261,29 +262,31 @@ func (r *Reconciler) setupSubscriptionStatusWatcher(ctx context.Context, ch *v1b
 				Ready:              corev1.ConditionTrue,
 			})
+		} else {
+			reconciled = false
```
realistically, if we know that we haven't properly reconciled here, why are we even continuing down the rest of the function?

lberk commented:
https://github.com/knative-sandbox/eventing-kafka/blob/3784fe05d38212764c6130f3db0b7d61b8bb4c29/pkg/channel/consolidated/status/status.go#L186-L221
tbh I'm finding this lambda block really hard to review now.  I'd appreciate it if you could abstract this into its own function with named return variables so I can understand the logic a bit better.  Especially because now we're dealing with prematurely unlocking and re-locking.

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -182,9 +191,26 @@ func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel,
+					*/
+					return true, true
+				} else {
+					if target.PodIPs.Len() == state.probedPods.Len() && target.PodIPs.HasAll(state.probedPods.List()...) {
```
would checking if the set is `Equal()` work?

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -182,9 +191,26 @@ func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel,
+						logger.Debugw("Dispatcher pods changed. Canceling old probes and restarting",
+							zap.Any("subscription", sub.UID))
+						// the dispatcher pods changed, we need to cancel old probing and start again
+						m.mu.Unlock()
```
why're we doing the unlocking, cancelprobing, re-lock when it seems like CancelProbing does the same thing as lines 214-216?

I'd prefer an approach where we're not unlocking, calling another function and then relocking like this tbh

devguyio commented:
/hold

devguyio committed:
Refresh pods probing on dispatcher pod changes

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -182,44 +186,25 @@ func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel,
-			// Cancel the polling for the outdated version
-			state.cancel()
-			delete(m.targetStates, subscriptionKey)
+		s, o := m.targetStates[sub.UID]
```
using single letter variable names for such extremely short lived variables is pretty idiomatic in go. I understand it's not readable if the variables would've had a bigger scope, but this is literally a single line scope where you'd know immediately what they from the assignment 3 lines above.

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -467,6 +509,15 @@ func (m *Prober) probeVerifier(item *workItem) prober.Verifier {
+func isOutdatedTargetState(s *targetState, sub eventingduckv1.SubscriberSpec, podIPs sets.String) bool {
+	s.readyLock.RLock()
+	defer s.readyLock.RUnlock()
+	r := s.ready
```
is adding a single use, single char variable like this for just the next line really more readable than just using `s.ready` ?

devguyio replied:
nope, that function changed a lot and that one was overlooked

devguyio commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -182,44 +186,25 @@ func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel,
-			delete(m.targetStates, subscriptionKey)
+		s, o := m.targetStates[sub.UID]
+		return s, o
+	}()
```
because of the lock, accessing the map needs to happen under lock and this construct of lock, defer ulock, inside an anonymous function is both idiomatic and less bug prone (since unlocking is deferred)

devguyio replied:
the whole idea is to rely on go's defer reliability which the go runtime will ensure to take place even if the call panics, something that the normal call to 'unlock' doesn't guarantee. Check the example by the end here https://blog.golang.org/defer-panic-and-recover

matzew replied:
perhaps adding a comment, on why?

the debate here shows, there is reasons, but also questions.  Comments usually help to get larger context

devguyio replied:
sure, will add a comment

lberk commented:
there's quite a few new code paths that unit testing isn't covering.  were there any planned additions to unit tests?

devguyio commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -354,6 +343,37 @@ func (m *Prober) CancelPodProbing(pod corev1.Pod) {
+	for _, state := range m.targetStates {
+		if !m.checkReadiness(state) {
+			// This is an in-flight state
+			sub := state.sub
```
the one on line 365 was overlooked.

devguyio replied:
I changed the one on 365.

devguyio commented:
@lberk yes and I created this issue to increase the unit tests coverage of the prober #491

devguyio commented:
/retest

devguyio commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -168,11 +170,13 @@ func (m *Prober) checkReadiness(state *targetState) bool {
 	)
-	return state.readyPartitions.Len() == int(partitions)
+	if !state.ready {
+		state.ready = state.readyPartitions.Len() == int(partitions)
```
`Len()` return type is `int`

devguyio committed:
change some bad var names

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio commented:
@lberk if you've done reviewing, do you mind giving me a summary and what does this PR needs to be merged from your POV?

devguyio commented:
/unhold

slinkydeveloper reviewed (commented):
The idea sounds correct to me, but I'll let somebody else approve it :)

lberk commented:
> @lberk yes and I created this issue to increase the unit tests coverage of the prober #491

So basically, no.  There will be no futher unit test additions with this PR.

lberk commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -157,9 +157,18 @@ func NewController(
 				pod, ok := obj.(*corev1.Pod)
 				if ok && pod != nil {
-					logger.Debugw("Dispatcher pod deleted. Canceling pod probing.",
+					logger.Debugw("Dispatcher pod deleted. Refreshing pod probing.",
```
If these two functions are essentially copy and pasted (save for a debug line).  Abstract the identical code into a different function and pass it with the debug string.

matzew replied:
I agree here, that `DeleteFunc` and `Addfunc` are very very identical - let's not start copy/paste here for no reason

matzew commented:
> So basically, no. There will be no futher unit test additions with this PR.

@lberk I see where you coming from, and I think we should always try have coverage of code via unit tests. 

I lack context of why the unit tests for the `status.go` are part of a different issue/PR and can not be part of this PR...

But I am fine in adding that afterwards, FWIW.

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -317,6 +320,11 @@ func (m *Prober) Start(done <-chan struct{}) chan struct{} {
+}
+
+// This is meant to be called while m.mu is locked
+func (m *Prober) forgetState(sub eventingduckv1.SubscriberSpec) {
```
could you elaborate to me why it's a sound design decision to have:
`CancelProbing` (which is part of the public Manager interface)
`CancelProbing`'s single job is to aquire the mutex lock, and call `forgetState(sub)`
`forgetState(sub)` has a comment a comment that indicates that we *must* have the lock before calling it
we then have other ways of calling `forgetState` elsewhere in the file.

Why aren't we using `CancelProbing` as the proper entrypoint my consistently?

matzew replied:
OK, so the issue is:

for outside invocation he have the `CancelProbe`, that basically does the lock and calls the private one.

Internally we call the `forget` since we have already taken care of the lock.

Perhaps the different names of `CancelProbe` versus `froget...` is confusing?

I think I understand the way of decorating the invocation of the `forget` with a lock, inside of the `CancelProbe`.

So...  Let's perhaps change the name?  `internalCancelProbe`, and be more explicit that it's only used w/in a mutex, and outside users should be instead calling `CancelProbe` ? 

And perhaps add a TODO... to have that sorted later

devguyio commented:
> So basically, no. There will be no futher unit test additions with this PR.

@lberk I misunderstood what you meant earlier by "were there any planned additions to unit tests?" I thought you meant "do you plan on adding unit tests later after this PR". And indeed I did, so I went ahead and created the issue to confirm that there's an issue and planned work for adding unit tests.

> I lack context of why the unit tests for the status.go are part of a different issue/PR and can not be part of this PR...

@matzew Reasons are: 

1. Time pressure.
2. We have some e2e tests in place that're giving us some level confidence.

lberk commented:
> @matzew Reasons are:
> 
>     1. Time pressure.
>     2. We have some e2e tests in place that're giving us some level confidence.

Neither of these are valid criterion.

devguyio commented:
/hold

devguyio commented:
@lberk I put it on hold to add unit tests and increase the coverage. Please let me know what else is needed for the PR to be approved? I'm addressing some comments too and responded to the others.

lberk commented:
https://github.com/knative-sandbox/eventing-kafka/blob/9128624c5e5a7d0f4f684585bb25e98b8a909bb8/pkg/channel/consolidated/status/status.go#L110
is that ingress line still valid here?

devguyio committed:
Add prober test for succeeding only after refreshing pod probing

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

devguyio commented:
> is that ingress line still valid here?
Nope, it was overlooked

devguyio commented:
@lberk @matzew I increased the test coverage , please check.
/unhold

devguyio commented:
Checking the failed tests

devguyio committed:
Update struct creation after unused field removed

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

lberk commented on pkg/channel/consolidated/status/status.go:
```diff
@@ -361,15 +361,13 @@ func (m *Prober) RefreshPodProbing(ctx context.Context) {
 					zap.Any("subscription", sub.UID))
 				return
 			}
-			if !state.probedPods.Equal(target.PodIPs) {
```
just so I understand, why was this condition no longer needed?

devguyio replied:
To force refreshing the Pods regardless of whether the IPs changed or not. We can't for sure know if pods with the same IPs are the same ones or not.

devguyio committed:
Reuse same handler for pod informer

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

lberk commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -154,26 +154,22 @@ func NewController(
+	return func(obj interface{}) {
+		pod, ok := obj.(*corev1.Pod)
+		if ok && pod != nil {
+			logger.Debugw("%s rods. Refreshing pod probing.", handlerType,
```
:) small typo

lberk commented on pkg/channel/consolidated/reconciler/controller/controller.go:
```diff
@@ -19,16 +19,15 @@ package controller
 
 	"knative.dev/eventing-kafka/pkg/apis/messaging/v1beta1"
 	"knative.dev/eventing-kafka/pkg/channel/consolidated/status"
+	kafkavmessaging1beta1 "knative.dev/eventing-kafka/pkg/client/informers/externalversions/messaging/v1beta1"
```
kafkavmessaging1beta1 or kafkamessagingv1beta1

devguyio replied:
woops, typo. Good catch

devguyio committed:
Fix a couple of typos

Signed-off-by: Ahmed Abdalla <aabdelre@redhat.com>

lberk commented on pkg/channel/consolidated/status/status_test.go:
```diff
@@ -39,110 +39,169 @@ import (
+	prober, ready := getProber(t, &lister)
+
 	done := make(chan struct{})
 	cancelled := prober.Start(done)
```
hm, just wondering out loud here, should we be `defer close(probeRequestsDX` elsewhere as well?

devguyio replied:
It wouldn't hurt, but it's not needed IMHO. The reason those channels are closed is to shutdown the prober and free the polling goroutines in the prober and not leak those routines. While closing the probeRequests won't be necessary since we're closing the whole server anyway and that'll clean up their goroutines IIUC. And as you know, you don't need to close all the channels since they'll be garbage collected if unreachable.

If you feel I should do that, sure I can. Let me know.

lberk commented:
/lgtm

Thanks for adding the tests @devguyio

devguyio commented:
@lberk sure, np. Are you gonna wait for someone else to approve the PR?

devguyio commented:
/woof

matzew reviewed (approved):
/lgtm
/approve

devguyio commented:
/woof
